### ISSUE TITLE
[Feature] Throw exception instead of returning false

### ISSUE THREAD

**Issue by dontub** [DirectoryTree/LdapRecord#186]
Currently some methods might return false if an LDAP operation fails. Depending on the problem it's also possible that an exception is thrown, though. This is inconsistent. Currently I have to catch exceptions and to check the return value to be sure that the operation succeeds. Instead of returning false an exception that contains the failed LDAP operation the LDAP error message and the LDAP error code as exception code should be thrown. To be BC this could be configurable. Though in my opinion returning booleans to report success should be dropped long term.

**Comment by stevebauman** [comment]
I totally agree with this. It's messy in its current form and you shouldn't have to do this when performing operations.

Worst case, this will be in v2.0. Best case we have it a configurable flag in v1 (like you mention) and then remove the flag in v2.

If you have time, I'd gladly accept a PR for the configurable option. If not, it may take a short bit for me to get this implemented 👍

**Comment by stevebauman** [comment]
After some thought and some digging, I've found Laravel handles this in the exact same way (returns booleans and also throws exceptions on SQL failure):

https://github.com/laravel/framework/blob/dd80e3867881a5644eaf5ec93cbba47dcdd0ba4e/src/Illuminate/Database/Eloquent/Model.php#L697-L703

Please provide a code example of an operation that fails silently from an LDAP server error and returns `false` so I can gather more insight for the usefulness of this option. Thanks!

**Comment by dontub** [comment]
Without looking at the PHP source code I don't know in which cases `false` is returned and in which cases it triggers an error which leads to an `LdapRecordException`
https://github.com/DirectoryTree/LdapRecord/blob/f3dfdbc647bfac9c04b8bf0d3671f519dd846877/src/Ldap.php#L877
https://github.com/DirectoryTree/LdapRecord/blob/f3dfdbc647bfac9c04b8bf0d3671f519dd846877/src/Connection.php#L307

Thus for saving I'm forced to use the following code as long as a boolean is returned:
```
try {
    if (!$model->save()) {
        // handle error, though I don't know why the operation failed and whether the LDAP was modified or not
    }
} catch (LdapRecordException $e) {
    // handle error
} catch (\Exception $e) {
   // handle error
}
```

What I would like to do is:
```
try {
    $model->save();
    // some more use of LdapRecord
} catch (LdapException $e) {
    /*
     * It should be clear from the exception type that an LDAP operation failed. 
     * I assume LdapRecordException is meant to be a base class for all exceptions
     * in the lib. Maybe it makes sense to use different exception types for the
     * different LDAP operation.
     */
     
    // handle error
    
    // LDAP error code
    $e->getCode();
    // LDAP error message and maybe additional information
    $e->getMessage();
    // The failed LDAP operation
    $e->getOperation();
    // The affected DN if the failed operation operates on a specific DN
    $e->getDN();
    // For operations modifying the LDAP argument $entry of the ldap_ function
    $e->getEntry();
} catch (\Exception $e) {
    // handle non LDAP error
}
```

As a workaround I've extended `Ldap` in this way:
```
class ExtendedLdap extends Ldap
{
    protected function executeFailableOperation(\Closure $operation)
    {
        $result = parent::executeFailableOperation($operation);
        if (false === $result) {
            $backtrace = debug_backtrace();
            $ldapOperation = $backtrace[1]['function'];
            $detailedError = $this->getDetailedError();
            if (!$this->isOkToContinue($ldapOperation, $detailedError->getErrorCode())) {
                throw (new LdapRecordException(
                    sprintf('LDAP %s failed: %s', $ldapOperation, $detailedError->getErrorMessage()),
                    $detailedError->getErrorCode()
                ))->setDetailedError($detailedError);
            }
        }

        return $result;
    }

    private function isOkToContinue(string $ldapOperation, int $errorCode): bool
    {
        return in_array($errorCode, [LdapResultCode::NO_SUCH_OBJECT, LdapResultCode::NO_RESULTS_RETURNED, LdapResultCode::MORE_RESULTS_TO_RETURN], true) &&
            ('read' === $ldapOperation || 'search' === $ldapOperation);
    }
}
```

**Comment by stevebauman** [comment]
Thanks @dontub I appreciate it.

Can you provide an example of any operation currently that returns `false` and does not generate an exception?

I've never experienced it -- as the only errors that won't be converted to exceptions are:

- `No server controls in result`
- `Partial search results returned`
- `Size limit exceeded`
- `No such object`

In your example, I'm not sure exactly what has been done to the `$model` instance which leads to this issue. For example, are you creating a user, resetting a password, adding group memberships, etc.

**Comment by dontub** [comment]
The PHP documentation doesn't define in which cases an error is triggered and in which it is not. So in case an LDAP operation fails an error MIGHT be triggered, but it does NOT HAVE TO be triggered. So if an `ldap_` function returns false an exception should be thrown. (I know that there are `openssl_` functions that in some cases just return `false`, but don't trigger an error.)

Looking at the code, you check the return value of the LDAP operation:
https://github.com/DirectoryTree/LdapRecord/blob/5d239cb7ca0aec218cf05c2b91fddb25672053cf/src/Models/Model.php#L977
Why, if `false` cannot be returned?

Nevertheless as long as `save()` returns a boolean indicating success I'm forced to check it. If it is sure that `false` cannot be returned than nothing at all should be returned. (Or a boolean that indicates if something was written to LDAP or no attribute in the model was changed.)

Slightly related: At some points you check if the error message contains some content. Wouldn't it be better and less implementation dependent to check the error code?

**Comment by stevebauman** [comment]
Hi @dontub,

I've now completed implementing this. It will be released in v2.0.0.

> Slightly related: At some points you check if the error message contains some content. Wouldn't it be better and less implementation dependent to check the error code?

This is simply my preference. I find it easier to read and understand what errors are being bypassed by seeing the error message itself, rather than using cryptic error codes. 👍
